File: app/jobs/crm_sync_job.rb
class CrmSyncJob < ApplicationJob
  queue_as :default
  retry_on StandardError, wait: :exponentially_longer, attempts: 3

  # Don't retry if contact was deleted
  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("[CrmSyncJob] Record not found: #{exception.message}")
  end

  def perform(contact_id, crm_type = nil)
    contact = Contact.find(contact_id)
    credentials = TwilioCredential.current

    # Guard: skip if no credentials or CRM sync disabled for contact
    return unless credentials
    return unless contact.crm_sync_enabled

    results = {}

    # Sync to Salesforce
    if (crm_type.nil? || crm_type == 'salesforce') && credentials.enable_salesforce_sync
      service = CrmSync::SalesforceService.new(contact)
      results[:salesforce] = service.sync_to_salesforce
    end

    # Sync to HubSpot
    if (crm_type.nil? || crm_type == 'hubspot') && credentials.enable_hubspot_sync
      service = CrmSync::HubspotService.new(contact)
      results[:hubspot] = service.sync_to_hubspot
    end

    # Sync to Pipedrive
    if (crm_type.nil? || crm_type == 'pipedrive') && credentials.enable_pipedrive_sync
      service = CrmSync::PipedriveService.new(contact)
      results[:pipedrive] = service.sync_to_pipedrive
    end

    Rails.logger.info "CRM sync completed for contact #{contact_id}: #{results}"
    results
  rescue StandardError => e
    Rails.logger.error "CRM sync job failed for contact #{contact_id}: #{e.message}"
    raise
  end
end

File: laravel-app/app/Jobs/CrmSyncJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Services\CrmSync\SalesforceService;
use App\Services\CrmSync\HubspotService;
use App\Services\CrmSync\PipedriveService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class CrmSyncJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;
    public $crmType;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 3;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30, 90];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 120;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     * @param string|null $crmType
     */
    public function __construct(int $contactId, $crmType = null)
    {
        $this->contactId = $contactId;
        $this->crmType = $crmType;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return array
     */
    public function handle()
    {
        try {
            $contact = Contact::findOrFail($this->contactId);
            $credentials = TwilioCredential::current();

            // Guard: skip if no credentials or CRM sync disabled for contact
            if (!$credentials || !$contact->crm_sync_enabled) {
                return [];
            }

            $results = [];

            // Sync to Salesforce
            if ((is_null($this->crmType) || $this->crmType === 'salesforce') && $credentials->enable_salesforce_sync) {
                $service = new SalesforceService($contact);
                $results['salesforce'] = $service->syncToSalesforce();
            }

            // Sync to HubSpot
            if ((is_null($this->crmType) || $this->crmType === 'hubspot') && $credentials->enable_hubspot_sync) {
                $service = new HubspotService($contact);
                $results['hubspot'] = $service->syncToHubspot();
            }

            // Sync to Pipedrive
            if ((is_null($this->crmType) || $this->crmType === 'pipedrive') && $credentials->enable_pipedrive_sync) {
                $service = new PipedriveService($contact);
                $results['pipedrive'] = $service->syncToPipedrive();
            }

            Log::info("CRM sync completed for contact {$this->contactId}: " . json_encode($results));
            return $results;

        } catch (Exception $e) {
            Log::error("CRM sync job failed for contact {$this->contactId}: {$e->getMessage()}");
            throw $e;
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::error("[CrmSyncJob] Failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
